[MLIR][OpenMP] Fix type mismatch in linear clause for INTEGER(8) variables

KrxGu opened the pull request:
Fixes #173332 

The compiler was crashing when compiling OpenMP `parallel do simd` with a `linear` clause on `INTEGER(8)` variables. The assertion failure occurred during MLIR-to-LLVM translation:
Cannot create binary operator with two operands of differing type!

**Root Cause:**
The bug was in `LinearClauseProcessor::updateLinearVar()` where the step value (i32) and induction variable were multiplied without normalizing to the linear variable's type (i64), causing type mismatches in LLVM IR generation.

**Solution:**
Updated the translation logic to cast both the induction variable and step value to `linearVarTypes[index]` before performing arithmetic operations. This ensures type consistency for both integer and floating-point linear variables.

**Testing:**
- Added integration test verifying successful compilation to LLVM IR
- Added lowering test for MLIR generation with various linear clause forms
- Verified the exact reproducer from the issue now compiles without errors

KrxGu committed:
[MLIR][OpenMP] Fix type mismatch in linear clause with INTEGER(8) variables

Fixes #173332 where compiling OpenMP parallel do simd with a linear
clause on INTEGER(8) variables crashed with assertion failure. The
issue was in MLIR-to-LLVM translation where step and induction variable
types weren't normalized to the linear variable's type before arithmetic
operations.

The fix ensures all mul/add operations use consistent types by casting
both the induction variable and step to match linearVarTypes[index].

KrxGu committed:
Address review feedback and fix test patterns

- Add error handling for invalid linear variable types
- Fix FileCheck patterns to use func.func instead of func
- Simplify test expectations to be more robust
- Use CHECK-DAG where order doesn't matter

KrxGu committed:
Fix integration test to prevent dead code elimination

- Initialize variable j and use it in the loop body
- Update CHECK-LABEL to match actual function name (repro_issue_173332_)
- Ensures the compiler generates IR for the test case

KrxGu commented:
@k-arrows this PR is ready for review.

CI is all green, Good To Merge!!

KrxGu committed:
Address reviewer feedback

- Remove unnecessary helper lambda (CreateSExtOrTrunc already handles same-type case)
- Move integer type check outside if-else for consistency across both paths

KrxGu commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -187,17 +187,45 @@ class LinearClauseProcessor {
+      llvm::Value *iv = loopInductionVar;
+      llvm::Value *step = linearSteps[index];
+
+      // Helper to cast signed integers with sign extension or truncation
```
Gotcha!

KrxGu commented:
@kparzysz i have addressed all the reviews. 

CI is all green, Good To Merge!!
Happy New year!!

NimishMishra commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -187,17 +187,37 @@ class LinearClauseProcessor {
+
+      if (linearVarType->isIntegerTy()) {
+        // Integer path: normalize all arithmetic to linearVarType
+        iv = builder.CreateSExtOrTrunc(iv, linearVarType);
```
Would a single `step = builder.CreateSExtOrTrunc(step, iv->getType());` not work here, like the Float path below?

NimishMishra commented on flang/test/Lower/OpenMP/parallel-do-simd-linear-i8.f90:
```diff
@@ -0,0 +1,79 @@
+    ! CHECK-DAG: %[[J:.*]]:2 = hlfir.declare {{.*}} {uniq_name = "_QFtest_multiple_i8Ej"}
+    ! CHECK-DAG: %[[K:.*]]:2 = hlfir.declare {{.*}} {uniq_name = "_QFtest_multiple_i8Ek"}
+    ! CHECK-DAG: %[[STEP:.*]] = arith.constant 1 : i32
+    ! CHECK: omp.simd linear(%[[J]]#0 = %[[STEP]] : !fir.ref<i64>, %[[K]]#0 = %[[STEP]] : !fir.ref<i64>)
```
Is there something specific you are checking during FIR lowering? Since the truncation / sign-ext is happening during translation of the MLIR op, it is best to have MLIR tests. I think we can do away with this test explicitly.

NimishMishra reviewed (commented):
Thanks for the fix. A couple of comments.

tblah reviewed (commented):
Thanks for the bug fix!

KrxGu committed:
Replace Fortran tests with direct MLIR translation test

- Remove flang integration and lowering tests per reviewer feedback
- Add MLIR test in openmp-llvm.mlir that directly tests the translation
- Test verifies i32 step with i64 linear variable generates correct sext
- Test confirms mul and add operations use consistent i64 types

Addresses review comments from NimishMishra and tblah

NimishMishra reviewed (commented):
Thanks for the changes. It looks okay to me. Just one request: is it also possible to add a test for the float path too? Just to check the `else` part too.

KrxGu committed:
Add test for floating-point linear variables in SIMD loops

Per reviewer feedback from NimishMishra, added test case for the
floating-point path in updateLinearVar to verify that i32 step values
are correctly handled with f64 linear variables.

NimishMishra reviewed (approved):
LGTM, thanks. Please wait for a few days for others to take a look. But please do land this before LLVM 22 branches.
